iceoryx2/tests/sample_mut_tests.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#[generic_tests::define]
+mod sample_mut {
+    use iceoryx2::port::publisher::Publisher;
+    use iceoryx2::port::subscriber::Subscriber;
+    use iceoryx2::prelude::*;
+    use iceoryx2::service::builder::publish_subscribe::PublishSubscribeCreateError;
+    use iceoryx2::service::port_factory::publish_subscribe::PortFactory;
+    use iceoryx2::service::Service;
+    use iceoryx2_bb_posix::unique_system_id::UniqueSystemId;
+    use iceoryx2_bb_testing::assert_that;
+
+    const MAX_LOANED_SAMPLES: usize = 5;
+
+    fn generate_name() -> ServiceName {
+        ServiceName::new(&format!(
+            "service_tests_{}",
+            UniqueSystemId::new().unwrap().value()
+        ))
+        .unwrap()
+    }
+
+    struct TestFixture<Sut: Service> {
+        service_name: ServiceName,
+        service: PortFactory<Sut, u64>,
+        publisher: Publisher<Sut, u64>,
+        subscriber: Subscriber<Sut, u64>,
+    }
+
+    impl<Sut: Service> TestFixture<Sut> {
+        fn new() -> Self {
+            let service_name = generate_name();
+            let service = Sut::new(&service_name)
+                .publish_subscribe()
+                .max_publishers(1)
+                .create::<u64>()
+                .unwrap();
+
+            let publisher = service
+                .publisher()
+                .max_loaned_samples(MAX_LOANED_SAMPLES)
+                .create()
+                .unwrap();
+
+            let subscriber = service.subscriber().create().unwrap();
+
+            Self {
+                service_name,
+                service,
+                publisher,
+                subscriber,
+            }
+        }
+    }
+
+    #[test]
+    fn when_going_out_of_scope_it_is_released<Sut: Service>() {
+        let test = TestFixture::<Sut>::new();
+
+        let mut sample_vec = vec![];
+
+        for _ in 0..4 {
+            while let Ok(sample) = test.publisher.loan() {
+                sample_vec.push(sample);
+            }
+
+            assert_that!(sample_vec, len MAX_LOANED_SAMPLES);
+            sample_vec.clear();
+        }
+    }
```

Comment:
I will make the test more expressive. I will confirm that no more samples can be acquired with loan and after `sample_vec.clear()` was called it works again.

In the current test the samples go out of scope in `sample_vec.clear()` and it is confirmed that `loan` works again since `assert_that!(sample_vec, len MAX_LOANED_SAMPLES);`